### ISSUE TITLE
[proposal] The `ifs` function

### ISSUE THREAD

**Issue by DmitrySharabin** [mavoweb/mavo#806]
While working on [the Memory Game](https://dmitrysharabin.github.io/mavo-memory-game/), I needed to evaluate a player’s rating depending on the number of moves they perform (the fewer moves they do, the more stars they earn). I decided to go with the built-in `if` function and ended up with [a couple of nested ifs](https://github.com/DmitrySharabin/mavo-memory-game/blob/521230bbaf04d8ca9867706c6c7dd65b64b43a72/index.html#L55):

```html
if(move < 19, 3, if(move < 26, 2, if(move < 32, 1, 0)))
```

If I decided to go up with _a five-star rating_, I would end up with a very hard-to-read expression (I tried, it was painful 😅).

So I thought to myself, wouldn’t it be beneficial if we had a function that would let authors get a readable expression (without a bunch of nested `if`s) in cases like mine? And I recalled that spreadsheet users faced the same issue (_nested-ifs-hell_) for a long time. And recently Microsoft and Google added the `ifs` function to their spreadsheets apps: [Microsoft Excel](https://support.microsoft.com/en-us/office/ifs-function-36329a26-37b2-467c-972b-4a39bd951d45) and [Google Sheets](https://support.google.com/docs/answer/7014145?hl=en).

This function evaluates multiple conditions (tests) and returns a value that corresponds to the _first_ true condition (test). It has the following syntax:

```
ifs(test1, iftrue1, test2, iftrue2, …)
```

In Google Sheets, every _test_ can also be a list of values.

This function would let us transform this expression:

```
if(move < 19, 3, if(move < 26, 2, if(move < 32, 1, 0)))
```

into this one:

```
ifs(move < 19, 3, move < 26, 2, move < 32, 1, 0)
```

In this example, I used the fact that the `ifs` function lets authors _provide a default value_ if none of the provided conditions (tests) are met. The previous expression is shorthand for:

```
ifs(move < 19, 3, move < 26, 2, move < 32, 1, true, 0)
```

I have implemented a prototype for this function. You can find it [here](https://codepen.io/dmitrysharabin/pen/MWEOeVO?editors=1010).

What do you think if we add the `ifs` function to the Mavo core? Or should it be in a separate plugin?

**Comment by LeaVerou** [comment]
I've been wondering what's the best syntax for this for quite a while. I needed it recently for my [homesearch app](https://projects.verou.me/homesearch/) and ended up writing [a custom JS function](https://github.com/LeaVerou/homesearch/blob/main/index.js#L33) (though in the end that allowed me to have more complex logic, as you can see). 

My concern with your `ifs()` function is that it's unclear what's the condition and what's the value. This may be acceptable for only 1 condition, but for something like that with multiple conditions, you'd need to start counting arguments to figure out if the position is even or odd. It's quite awkward. I think it should be somehow grouped to make it obvious immediately what the conditions are and what the values are.

**Comment by karger** [comment]
I too have run into nested ifs.  And it would be appealing to have a general solution.  But is the common case something that might be addressed with a switch/case type expression, and might this have simpler syntax?

**Comment by karger** [comment]
OTOH, I think the concern about confusing conditions and values might not be severe, since conditions are usually clearly boolean expressions.

**Comment by karger** [comment]
you might also offer "else" as syntactic sugar for "true" in the final condition?

**Comment by karger** [comment]
As for dmitry's particular memory game application, I think the most elegant approach there would be to have our own version of vlookup.   lookup(key, values, outputs, true) could take a sorted list of values and equal length list of outputs, find in that values list the largest value smaller than the key, and return the item at the corresponding index of output.  The 4th "true" argument is what tells vlookup in excel to do an "approximate search", ie look for the closest value instead of an exact match.

**Comment by LeaVerou** [comment]
*(@karger, I merged your consecutive comments)*

The problem with a switch statement is that the use cases often need a range, so it would be nice if the syntax allowed for that too (perhaps via an array?)

**Comment by karger** [comment]
Right, that limitation of the switch statement is why I suggested approximate vlookup.   https://www.excel-university.com/perform-approximate-match-and-fuzzy-lookup-in-excel/

**Comment by DmitrySharabin** [comment]
If the author needs any functions from Microsoft Excel or Google Sheets, they might include the [FormulaJS](https://github.com/formulajs/formulajs) library in their app. Mavo handles all the functions supported by the library perfectly.

Here is a tiny demo: https://codepen.io/dmitrysharabin/pen/MWxWqGj

**Comment by LeaVerou** [comment]
I actually quite like [Coda’s `switchIf()` function](https://coda.io/formulas#SwitchIf), but since it's compatible with `if()`, not sure why it would even need to be a separate function: we could just optionally support additional arguments in `if()`, getting the best of both worlds.
Yes, it does still have the downside that it's hard to see what's the condition and what's the value without proper formatting, but maybe that's something we can live with.

**Comment by DmitrySharabin** [comment]
I believe the working prototype (https://codepen.io/dmitrysharabin/pen/MWEOeVO) shows that we already have everything we need: the `ifs` function works the same way as `if` (if arguments <= 3).

**Comment by LeaVerou** [comment]
My point is, why do we need a separate function at all?

**Comment by DmitrySharabin** [comment]
I probably badly expressed myself—I meant that we could replace the code of `if` with the code of `ifs` from the prototype, and it shouldn't break anything we already have and will augment the `if` function.

**Comment by LeaVerou** [comment]
Yes, let's!

Perhaps we can also ship it at a similar time as https://github.com/mavoweb/mavo/issues/587#issuecomment-1878664336 since the cost of executing everything if we have multiple conditions is even worse.
